src/executor_child.c

```diff
@@ -0,0 +1,149 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   executor_child.c                                   :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: sqiu <sqiu@student.42vienna.com>           +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/07/31 14:24:49 by sqiu              #+#    #+#             */
+/*   Updated: 2023/08/07 17:53:51 by sqiu             ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+/**
+ * @file executor_child.c
+ * @brief Behaviour of child processes.
+ */
+
+#include "mod_executor.h"
+
+/**
+ * @brief First cmd of pipe.
+ * 		fd_pipe[0] refers to the read end of the pipe.
+ * 		fd_pipe[1] refers to the write end of the pipe.
+ * Close all unrequired file descriptors.
+ * Assign fds to stdin and stdout.
+ * 		If fd_out exists, assign stdout with fd_out, else with fd_pipe[1]
+ * 		if fd_in exists, assign stdin with fd_in, else with 0
+ * Close assigned fds.
+ * Execute cmd.
+ * @param cmd 		Current cmd being processed.
+ * @param envp		String array of env.
+ * @param data		Data struct containing the env.
+ * @param builtin	Bool indicating if cmd is a builtin.
+ */
+void	ft_firstborn(t_cmd *cmd, char **envp, t_data *data, bool builtin)
+{
+	ft_close(cmd->fd_pipe[0]);
+	if (cmd->fd_out >= 0)
+	{
+		if (cmd->fd_in >= 0)
+			ft_replace_fd(cmd->fd_in, cmd->fd_out);
+		else
+			ft_replace_fd(0, cmd->fd_out);
+		ft_close(cmd->fd_out);
+	}
+	else
+	{
+		if (cmd->fd_in >= 0)
+			ft_replace_fd(cmd->fd_in, cmd->fd_pipe[1]);
+		else
+			ft_replace_fd(0, cmd->fd_pipe[1]);
+		ft_close(cmd->fd_pipe[1]);
+	}
```

Comment:
The devil lies in the detail. Every function duplicates different fds.

The errors are not retrieved on purpose as they are created in the child process and I don't see how they can be returned to the parent. The parent process will only get the exit status via pid, any other information of the child will be lost.

Reply:
Even if they dupl different: you can pass different vars, depending on what -born you are. 
Since it's more of a refactor story, we can leave it as is.

Except on question: ft_lastborn closes fd_out in any case, outside of the if condition. first and middle only if fd_out is >= 0 (which seems more logical to me). Is this an oversight?

The errors can be retrieved via the exit status of the child. Even if it's just a 1 instead of 0 it would indicate an error.

Reply:
The closing of fd_out may be moved inside the if condition fd_out >= 0. Technically it doesn't matter as ft_close() contains a check whether the fd rly needs to be closed and doesn't if fd was not set. But just for the beauty of uniformity it is now inside the if cond.